### PR TITLE
feat(appellate): Upload free documents Apellate PACER

### PR DESCRIPTION
**Pull request by ERosendo**
This PR adds support to upload free documents from Appellate PACER. This PR resolves https://github.com/freelawproject/recap/issues/322.

This PR introduces the following changes:

 - This PR adds three helper methods to get data related to document links:
     - `getDocNumberFromAnchor`: This helper method gets the **document number** from the anchor element that has a document link.
     - `getAttachmentNumberFromAnchor`: This utility function accesses the row that contains the document link and gets the **attachment number**.
     - `cleanDocLinkNumber`: this helper function removes whitespace from the **document number** and the **attachment number**.
 
- The `findDocLinksFromAnchors` helper method from the `appellate utils` is updated to add an `onClick` event to every document link. This event allows us to check if a **document link** redirects to a free document or to a **download confirmation page**.

- This commit also adds a few [data attributes](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes) to the anchor elements that have a document link ( `document number`, `attachment number`, `pacer_doc_id`, `case_id`). We need these data attributes to create the data object we send the API when the extension uploads files. 

- A new method called `onClickEventHandler` is added in the `appellate utils` file. This utility will make the findDocLinksFromAnchors method shorter and more readable.

- A new helper function called `handleFreeDocResponse` is added to the `pdf_upload.js` file. This helper checks the `content-type` header of the response. If the `content-type` is `application/pdf`, the extension gets the PDF file as a BLOB and starts the uploading process.

- We need the `document number` to upload files from **attachment pages** but this value is not on the page, so this PR adds this value as a parameter called `docNum` in the query string of each link.

- A new message is added to the background script to upload the free documents.

This approach to upload free documents works in the docket reports and the attachment page. Here's a gif showing the new feature:

Appellate-free-doc-attachment

### Additional comment 

- This PR also adds a check in the `case selection page` handler to avoid errors when the district court link is not available on the page. The current implementation of the extension fails to handle **case selection pages** if they don't have a district court link like docket [19-15077](https://www.courtlistener.com/docket/66680340/karastan-edwards-v-us-attorney-general/). Here's a screenshot of this case:

image

The table shows the **originating case number** but there isn't an anchor element that links to the district court.

### CODE REVIEW

**Review by mlissner — CHANGES_REQUESTED** [review]
A couple little comments, but this seems to work pretty well. A part of me feels like it's overly complicated. Can you write a list of how it works at a high level? Something like:

 - Some data attributes are set first, so they can be grabbed by a listener.
 - The listener is triggered by the click, and passes XYZ
 - The onlick sniffs the content type and ABC

It just feels a bit like we're jumping all over the place and storing data in a few different places to make this work. It sort of _smells_ like it could be better?

I did some testing and it seems to mostly work. I had one upload that didn't get associated with the docket properly in CL. Not sure why and haven't tracked it down, but it generally works, which feels pretty OK to me for a release, I think. We'll have bug reports, but that's OK.

File: src/appellate/utils.js [hunk 1]
```diff
@@ -177,42 +177,77 @@ let APPELLATE = {
...
-        let form = document.getElementsByName('doDocPostURLForm')[0];
-        if (form) {
-          form.dls_id.value = params.doc_id;
-          form.caseId.value = params.case_id;
-          form.submit();
-        }
-      });
-
-      let docId = PACER.getDocumentIdFromUrl(a.href);
+      clonedNode.onclick = function (e) {
```
[the remark this answers is not in the document]
Comment (ERosendo) [reply src/appellate/utils.js:232]: You're right about the comment and the helper's name.  Im going to change that.

Thanks.

File: src/pacer.js [hunk 2]
```diff
@@ -585,6 +585,78 @@ let PACER = {
             </div>`;
   },
 
+  // removes whitespace and returns a number
```
[the remark this answers is not in the document]
Comment (ERosendo) [reply src/pacer.js:588]: Yes. I will add test for this helper functions.

**Comment by ERosendo** [comment]
Sure. This new feature uses the following steps to check if a document is free and then uploads it to the RECAP archive:

1. The extension parses the `doc_id` and the `case_id` from the original **onClick** event handler of the `<a>` element.

1. Removes the original **onClick** attribute and updates the `href` attribute of the anchor:

   1. The `caseId` parameter is added to the `href` to fix [#324](https://github.com/freelawproject/recap/issues/324). We get this value in the first step.
   2. The `docNum` parameter is added to the `href` , so we can use it in pages like the **attachment menu** ( there's no way to get this value from this page). We get this value using the `getDocNumberFromAnchor` or we can use the **queryParameters** of the page (We use this approach in the **attachment page**. The extension should add the `docNum` when it reaches the **docket report**, so we can use it in the **attachment menu**).
  
1. The extension replaces the anchor element with a clone. We need this step to remove the default listener that PACER adds to these links ( The second step is not enough to remove the default listener). 

1. We add some data to the anchor elements using data attributes. Here's the list of the attributes we add:

   - `pacer_doc_id`: this is the normalized `doc_id` ( as explained [here](https://www.courtlistener.com/help/api/rest/#processing-queue-endpoint))
   - `pacer_dls_id`: this is the original `doc_id.` We got this value in the first step.
   - `pacer_case_id`: We get this value from the first step ( in **dockets reports**) or we can use the query parameters from the page URL( we use this approach in **attachment pages**).
   - `pacer_tab_id`: We get this value from the background script.
   - `document_number`: We get this value in the second step.
   - `attachment_number`: We get this value using the `getAttachmentNumberFromAnchor` method. 

1. The extension add a new **onClick** listener to every doc link. This listener helps us identify the free documents.

   1. When a user clicks one of these doc links, the event handler fires a `POST` request to `/TransportRoom` and uses some data attributes ( `pacer_dls_id` and  `pacer_case_id`) from the HTML element that was clicked ( the target element) to create a body of data to be sent in the request.
  
   3. We use the method called `handleFreeDocResponse` to check the `content-type` of the response.
       - if the content-type is not 'application/pdf', the method redirects the browser to a new page using the href attribute of the target element.
      
       - If the content-type is 'application/pdf', the method should complete the following steps
           -  saves the PDF file as a BLOB in the chrome storage ( we need the `pacer_tab_id` here )
           -  creates a new data object using some data attributes ( `attachment_number`, `document_number`, `pacer_case_id`, `pacer_doc_id`).
           - sends a message from the content script to the background worker to upload the file ( includes the data object in this message, so we can get all the values needed in the background script to complete the upload request).
           - redirects the user to a new page using the `href` attribute of the target.

**Comment by mlissner** [comment]
Thanks. This is helpful. Tomorrow, in the light of the new day, I'm curious if you see anything in the above that can be simplified. We don't necessarily have to do it now, maybe it's a refactor for later, but I think it's worth considering while it's all fresh in your mind. Let me know what you think — tomorrow, please! :)

**Comment by ERosendo** [comment]
I thought that **cloning** and **replacing** the element was enough to remove the old listener, so I tried to simplify step 2, but I noticed that if we don't remove the `onClick` attribute before we clone the element, the anchor keeps using JS code from PACER and doesn't fire the `onClick` event the extension creates.

After reviewing the steps, I realized the onClick event redirects the browser to a new page (using the `href` attribute of the target) and avoids opening new tabs, so we can remove this line: https://github.com/freelawproject/recap-chrome/blob/4c029f03d0d73d9b6ced813b2765423bdc5e76f0/src/appellate/utils.js#L193

### Additional comment

I noticed that the regex expression we use to parse data from **download confirmation pages** is not working properly. I was trying to replicate errors from https://github.com/freelawproject/recap/issues/292 and went to purchased docket 20-1422, then clicked docket entry 175 (which led to an **attachment page**) and I selected and purchased a document but the extension did not upload the file. I though It was related to the issue but I realized that the regex expression was failing.

We use the following regex expression to parse data from the **download page** of a document that belongs to an **attachment menu**. 

``` javascript
let dataFromAttachment = /^Document: PDF Document \(Case: ([^']*), Document: (\d)-(\d)\)/.exec(title_string); 
```

and works for strings like:

```javascript
"Document: PDF Document (Case: 20-15019, Document: 1-1)"
"Document: PDF Document (Case: 20-1422, Document: 2-1)"
```

but fails when the document number has more than one digit, like the following string:

```javascript
"Document: PDF Document (Case: 20-1422, Document: 175-2)"
```
Commit https://github.com/freelawproject/recap-chrome/pull/305/commits/1959abeb9de3029e50100c7eab9394929d971042 fixes this issue adding the repetition operator "+" in the expression

**Comment by mlissner** [comment]
I think it's this dance that makes me uncomfortable, but I'm not sure I know a better way:

 - Add a listener
 - Add data attributes
 - Add onlick
 - When clicked:
     - Run the onclick listener
     - gather the data attributes 
     - run a function from pdf_upload.js

That feels like a lot when all you really want to do is:

 - Run some code when something is clicked.

Maybe this is just the javascript way, I don't know, but it is a lot of misdirection to figure out:
 - where the code for the listener is when the link is clicked
 - where the listener is getting data attribtues
 - how those attributes got set in the first place, and 
 - where the handleFreeDocResponse function is (and what it does).

See what I mean?

**Review by mlissner — APPROVED** [review]

**Pull request merged**